tests/integration/test_katib_experiments.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Integration tests for Katib Experiments.
+"""
+
+import logging
+
+import lightkube
+import pytest
+from pytest_operator.plugin import OpsTest
+from utils import (
+    assert_deleted,
+    assert_exp_status_running_succeeded,
+    assert_get_experiment,
+    create_experiment,
+    delete_experiment,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.parametrize(
+    "experiment_file",
+    [
+        "tests/assets/crs/bayesian-optimization.yaml",
+        "tests/assets/crs/cmaes.yaml",
+        "tests/assets/crs/darts-cpu.yaml",
+        "tests/assets/crs/enas-cpu.yaml",
+        "tests/assets/crs/file-metrics-collector.yaml",
+        "tests/assets/crs/grid-example.yaml",
+        "tests/assets/crs/hyperband.yaml",
+        "tests/assets/crs/median-stop.yaml",
+        "tests/assets/crs/random.yaml",
+        "tests/assets/crs/simple-pbt.yaml",
+    ],
+)
+async def test_kaitb_experiments(ops_test: OpsTest, experiment_file):
+    """Test Katib Experiments.
+
+    Create experiment and assert that it is Running or Succeeded.
+    At the end, experiment is deleted.
+    NOTE: This test is re-using deployment created in test_deploy_katib_charms() in test_charms.py
+    """
+    namespace = ops_test.model_name
+    lightkube_client = lightkube.Client()
+
+    # Add metrics collector injection enabled label to namespace
+    test_namespace = lightkube_client.get(
+        res=lightkube.resources.core_v1.Namespace, name=namespace
+    )
+    test_namespace.metadata.labels.update(
+        {"katib.kubeflow.org/metrics-collector-injection": "enabled"}
+    )
+    lightkube_client.patch(
+        res=lightkube.resources.core_v1.Namespace,
+        name=test_namespace.metadata.name,
+        obj=test_namespace,
+    )
+
+    exp_name = create_experiment(
+        client=lightkube_client, exp_path=experiment_file, namespace=namespace
+    )
+
+    assert_get_experiment(logger, lightkube_client, exp_name, namespace)
+    assert_exp_status_running_succeeded(logger, lightkube_client, exp_name, namespace)
+
+    delete_experiment(lightkube_client, exp_name, namespace)
+    assert_deleted(logger, lightkube_client, exp_name, namespace)
+
+    # Wait for applications to settle
+    await ops_test.model.wait_for_idle(
+        status="active",
+        raise_on_blocked=True,
+        timeout=360,
+    )
```

Comment:
Perhaps I'm missing something here, but why would the application not be idle?

Reply:
from the things I've seen in GH runners, when resources are exhausted, some of the applications specifically `katib-controller` goes to error with:
```
juju status:
katib-controller  .../katib-controller:v0.15.0  error        1  katib-controller                          0  10.152.183.196  no       unknown container reason "ContainerStatusUnknown": The container could not be located when the pod was terminated
```

Reply:
This feels weird though, because our katib-controller shouldn't be affected by anything running as part of these tests. If there's a resource issue on the cluster I would expect the tests to fail because K8s has failed, not because a random app (which katib-controller in fact is here) did.

Reply:
I agree that it's weird but it was gone with increasing resources, you can see the logs [here](https://github.com/canonical/katib-operators/actions/runs/5810563846/job/15759335056)

Reply:
It's the Katib Controller (upstream) failing though, isn't it? Our charm goes into error state due to the failing healthcheck, but I understand that the tests would fail even without asserting that `katib-controller-operator` is healthy: https://github.com/canonical/katib-operators/actions/runs/5810563846/job/15759335056#step:4:345

Reply:
yes they would fail, but they'd fail earlier once the container is not up, I think it'd make sense if the check was a pre-check at the beginning of the test. how can we stop the test run once the charm is in error or things are not working as expected to be time efficient? is what I'm thinking

Reply:
If we can't check the status of the katib-controller container directly I'd say let's keep this, adding a comment that the charm is only expected to fail here because the katib-controller might die due to lack of resources (not because of an issue on our side).

I mean, I would probably prefer something like `ops_test.model.unit.get_container("katib-controller-whatever-the-name-is").can_connect()`, I think it's more straightforward.

Reply:
my only concern here is we're not sure what other containers might be affected by lack of resources, I see no harm in having it as a generic check to ensure things are up. The goal is just to avoid creating experiments in an unexpected environment.